path_dynamic_roles.go

```diff
@@ -0,0 +1,295 @@
+				},
+			},
+			HelpSynopsis:    "List all the dynamic roles Vault is currently managing in OpenLDAP.",
+			HelpDescription: "List all the dynamic roles being managed by Vault.",
+		},
+		{
+			// Pattern: dynamicCredPath + framework.GenericNameRegex("name"),
+			Pattern: path.Join(dynamicCredPath, framework.MatchAllRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the dynamic role.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicCredsRead,
+				},
+			},
+			HelpSynopsis: "Request LDAP credentials for a dynamic role. These credentials are " +
+				"created within OpenLDAP when querying this endpoint.",
+			HelpDescription: "This path requests new LDAP credentials for a certain dynamic role. " +
+				"The credentials are created within OpenLDAP based on the creation_ldif specified " +
+				"within the dynamic role configuration.",
+		},
+	}
+}
+
+func secretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: secretCredsType,
+		Fields: map[string]*framework.FieldSchema{
+			"username": {
+				Type:        framework.TypeString,
+				Description: "Username of the generated account",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password to access the generated account",
+			},
+			"DNs": {
+				Type:        framework.TypeStringSlice,
+				Description: "List of the DNs created",
+			},
+		},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *backend) pathDynamicRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("name").(string)
+
+	rawData := data.Raw
+	err := convertToDuration(rawData, "default_ttl", "max_ttl")
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert TTLs to duration: %w", err)
+	}
+
+	dRole := &dynamicRole{}
+	err = mapstructure.WeakDecode(rawData, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	if dRole.CreationLDIF == "" {
+		return nil, fmt.Errorf("missing creation_ldif")
+	}
+
+	dRole.CreationLDIF = decodeBase64(dRole.CreationLDIF)
+
+	err = assertValidLDIFTemplate(dRole.CreationLDIF)
+	if err != nil {
+		return nil, err
+	}
+
+	err = storeDynamicRole(ctx, req.Storage, roleName, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to save dynamic role: %w", err)
+	}
+
+	return nil, nil
+}
+
+// convertToDuration all keys in the data map into time.Duration objects. Keys not found in the map will be ignored
+func convertToDuration(data map[string]interface{}, keys ...string) error {
+	merr := new(multierror.Error)
+	for _, key := range keys {
+		val, exists := data[key]
+		if !exists {
+			continue
+		}
+
+		switch v := val.(type) {
```

Comment:
You could use `parseutil.ParseDurationSecond` here instead.

Reply:
Ah good catch. I forgot that we had that package. Updated